lib/azure/core/http/retry_policy.rb

```diff
@@ -47,8 +47,8 @@ def call(req, _next)
             @retry_data[:error] = $!
           end while should_retry?(response, @retry_data)
           # Assign the error when HTTP error is not thrown from the previous filter
-          @retry_data[:error] = response.error if response && !response.success?
-          if @retry_data.has_key?(:error)
+          if response && !response.success?
```

Comment:
Thanks for reporting the issue and raising PR.
This changes makes the error only be thrown when response has a value. But when the first time `response = _next.call` is called and failed, `response` is `nil` and `@retry_data[:error]` is assigned by `@retry_data[:error] = $!`. In this case, the error is not thrown.

To fix the issue, we may set `@retry_data[:error]` to `nil` every time before we call `response = _next.call`.